ER-413 Gov one login Integration

jack-coggin opened the pull request:
- [ER-413]
- [ER-869]
- [ER-867]

Add gov one login authentication option with one login integration environment. Also adds bridging page and t&c page and updates homepage and user account page.


The bridging page can be accessed at `/gov-one/info` of from the sign in links

Requires `GOV_ONE_LOGIN` env variable feature flag

<br />

For further information and technical checklist see `gov_one_login/GOV_ONE_LOGIN.MD`


[ER-413]: https://dfedigital.atlassian.net/browse/ER-413?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ER-869]: https://dfedigital.atlassian.net/browse/ER-869?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ER-867]: https://dfedigital.atlassian.net/browse/ER-867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

peterdavidhamilton commented on .gitignore:
```diff
@@ -57,6 +57,10 @@ localhost.crt
 nscacert.pem
 
+# gov one login keys
+private_key.pem
```
Are these in the credentials now?

jack-coggin replied:
Yes [7313839](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/7313839ff0455bbdda6aa8c474202632476b98ec)

peterdavidhamilton commented on db/migrate/20231031094611_add_id_token_to_users.rb:
```diff
@@ -0,0 +1,6 @@
+class AddIdTokenToUsers < ActiveRecord::Migration[7.1]
```
Migration version numbers are different for each file.

jack-coggin replied:
[48f4ae5](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/48f4ae55aa6a2b473b80efdb367a34189d887d8c)

peterdavidhamilton commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,88 @@
+      client_assertion: jwt_assertion,
+    }
+
+    token_uri = URI.parse("#{ENV['GOV_ONE_BASE_URI']}/token")
```
private method?

jack-coggin replied:
[7313839](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/7313839ff0455bbdda6aa8c474202632476b98ec)

peterdavidhamilton commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,88 @@
+
+  # @return [Hash]
+  def tokens
+    body = {
```
private method?

jack-coggin replied:
I call this method in the `omniauth_callbacks_controller`

peterdavidhamilton replied:
I was thinking how we could reduce method size, so extracting `body` into a private method for example. Would that be possible?

peterdavidhamilton commented:
One error is a timeout on the CMS, if that happens just rerun it. Another spec failed on a count of locale. As you may add some new ones perhaps add a quick inventory in the PR description so Content can create them.

jack-coggin commented on app/views/gov_one/info.html.slim:
```diff
@@ -0,0 +1,17 @@
+    = m('gov-one-info.body')
+
+    - if current_user
+      = govuk_button_link_to 'Sign out of Gov One Login', logout_uri
```
Sure [7313839](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/7313839ff0455bbdda6aa8c474202632476b98ec)

jack-coggin commented on app/models/user.rb:
```diff
@@ -21,12 +21,22 @@ def self.dashboard_headers
+  # @param gov_one_id [String]
+  # @return [User]
+  def self.create_from_gov_one(email:, gov_one_id:)
+    user = User.new(email: email, gov_one_id: gov_one_id, confirmed_at: Time.zone.now)
```
[7313839](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/7313839ff0455bbdda6aa8c474202632476b98ec)

peterdavidhamilton commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,95 @@
+    body = {
+      grant_type: 'authorization_code',
+      code: @code,
+      redirect_uri: ENV['GOV_ONE_REDIRECT_URI'],
```
New ENVs will want to be added to our example file which records all of them in one place. I believe this one should also be set for the dev composition, is that right?

jack-coggin replied:
[a83da4e](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/a83da4e7325fda7d29d4f24c8abac57759bdedd0) I added the env variables to the example

peterdavidhamilton commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,80 @@
+  end
+
+  # @return [User]
+  def find_or_create_user(email, id)
```
I feel like this method might come out a little cleaner if it was given some more attention.

jack-coggin replied:
Ah yes, that was messy! [773178d](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/773178de60d063b1e23b023cab9470aec05273c7)

peterdavidhamilton replied:
What would happen if we gave the responsibility of all this to a `User.find_or_create_from_gov_one` method?

peterdavidhamilton commented on app/helpers/application_helper.rb:
```diff
@@ -14,11 +14,11 @@ def navigation
         header.with_action_link(text: 'Sign out', href: destroy_user_session_path, options: { id: 'sign-out-desktop', data: { turbo_method: :get }, inverse: true })
```
Can you revert the changes in this file please

jack-coggin replied:
[773178d](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/773178de60d063b1e23b023cab9470aec05273c7)

peterdavidhamilton commented:
Is there any additional documentation, class-level comments, method input/output missing? 
Are we going to use a feature flag around this?

jack-coggin commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,98 @@
+class GovOneAuthService
+  # @param code [String]
+  def initialize(code)
```
[773178d](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/773178de60d063b1e23b023cab9470aec05273c7)

jack-coggin commented:
> Is there any additional documentation, class-level comments, method input/output missing? Are we going to use a feature flag around this?

@peterdavidhamilton I've updated the YARD documentation where I could see it was missing, added a few more comments and have added a feature flag env variable

jack-coggin commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,100 @@
+# Service for interacting with Gov One Login
```
[ceaa9dd](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/ceaa9ddff65713e337822b842d91ec7e04a75e75)

ebrett commented on config/initializers/devise.rb:
```diff
@@ -306,9 +306,21 @@ def i18n_message(default = nil)
+    client_options: {
+      port: 443,
+      scheme: 'https',
+      host: 'oidc.integration.account.gov.uk',
```
Should this be GOV_ONE_BASE_URI ?

jack-coggin replied:
Yes, thanks [7294ab7](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/7294ab73601bba2594c50ed4a4e2e664298b9593)

jack-coggin commented on spec/services/gov_one_auth_service_spec.rb:
```diff
@@ -0,0 +1,61 @@
+      it 'returns a hash of user info' do
+        allow(mock_response).to receive(:body).and_return(user_info_payload.to_json)
+
+        result = auth_service.user_info(access_token)
```
[d297b42](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/d297b42eaadec941d65376a5584f9f28b9e2ea87) ?

peterdavidhamilton replied:
That is one way to do it but how many other specs will use this new shared context? If it were to be shared it should be in the support folder incase you fire off an individual spec. I was suggesting standard `let` and `before` blocks.

jack-coggin commented on app/helpers/application_helper.rb:
```diff
@@ -1,28 +1,35 @@
   def navigation
-    render(HeaderComponent.new(classes: 'dfe-header noprint', container_classes: %w[dfe-header-f-header-flex], navigation_label: 'Primary navigation')) do |header|
-      header.with_navigation_item(text: 'Home', href: root_path, classes: %w[dfe-header__navigation-item])
+    govuk_header(classes: 'noprint') do |header|
```
No it doesn't - thanks

peterdavidhamilton commented on app/services/gov_one_auth_service.rb:
```diff
@@ -1,12 +1,17 @@
+# - decode an id token to get the user's gov one id
+#
+# More information on the Gov One Login integration environment can be found here:
+# https://docs.sign-in.service.gov.uk/integrate-with-integration-environment/
```
Do you like the look of it in YARD? There is a `@see` tag you could try on the link. We can chat about it.

jack-coggin replied:
Lovely!

peterdavidhamilton commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,100 @@
+
+  # @return [Hash]
+  def jwks
+    discovery_url = "#{ENV['GOV_ONE_BASE_URI']}/.well-known/jwks.json"
```
Let's talk about this suggestion even if you don't apply it because it might be useful.

peterdavidhamilton commented on config/application.rb:
```diff
@@ -115,12 +115,12 @@ def debug?
     # @return [Boolean]
     def gov_one_login?
-      Types::Params::Bool[ENV.fetch('GOV_ONE_LOGIN', false)]
+      Types::Params::Bool[ENV.fetch('GOV_ONE_LOGIN', true)] && !live?
```
So after this feature is merged, enabled in staging, but off in prod; what's the workflow going to be if we have 20 subsequent versions before we are ready to enable the feature?

peterdavidhamilton commented on app/views/gov_one/info.html.slim:
```diff
@@ -8,13 +8,7 @@
-    - else
-      = govuk_button_link_to t('gov-one-info.sign-in-button'), login_uri.to_s
-=======
       = govuk_button_link_to t('gov_one.links.sign_out'), logout_uri.to_s
```
I actually think it would make sense to make these two buttons into their own helper methods

peterdavidhamilton commented on app/models/user.rb:
```diff
@@ -21,12 +21,27 @@ def self.dashboard_headers
+  # @param email [String]
+  # @param gov_one_id [String]
+  # @return [User]
+  def self.find_or_create_from_gov_one(email:, gov_one_id:)
```
This doesn't yet have a unit test. Now all the post login logic is here the `User` object can be tested more easily.

peterdavidhamilton commented on docker-compose.dev.yml:
```diff
@@ -19,8 +19,9 @@ services:
       - DATABASE_URL=postgres://postgres:password@db:5432/early_years_foundation_recovery_development
       - RAILS_ENV=development
-      - DOMAIN=app:3000
+      - DOMAIN=localhost:3000 # put back to app
```
Can you put this back and see if things still work

peterdavidhamilton commented on docker-compose.dev.yml:
```diff
@@ -19,8 +19,9 @@ services:
-      - DOMAIN=app:3000
+      - DOMAIN=localhost:3000 # put back to app
       - RAILS_SERVE_STATIC_FILES=true
+      - GOV_ONE_BASE_URI=https://oidc.integration.account.gov.uk
```
👀

peterdavidhamilton commented on spec/services/gov_one_auth_service_spec.rb:
```diff
@@ -0,0 +1,106 @@
+  end
+
+  describe 'ENDPOINTS' do
+    it 'returns a frozen hash of endpoints urls' do
```
You can likely produce better documentation doing each one individually

```ruby
describe GovOneAuthService::ENDPOINTS do
  specify 'login' do
    expect(described_class[:login]).to eq 'https://oidc.test.account.gov.uk/authorize'
  end
end
```

peterdavidhamilton commented on spec/services/gov_one_auth_service_spec.rb:
```diff
@@ -0,0 +1,106 @@
+
+    it 'returns a hash of correct jwt payload' do
+      expect(jwt_payload[:aud]).to eq('https://oidc.test.account.gov.uk/token')
+      expect(jwt_payload[:iss]).to be_a(String)
```
Don't you control the value coming from `Rails.application.config.gov_one_client_id` and credentials. Without revealing sensitive info can we assert a real value? Does the `test` env need real creds?

peterdavidhamilton commented on app/helpers/gov_one_helper.rb:
```diff
@@ -0,0 +1,39 @@
+    gov_one_uri(:login, params)
+  end
+
+  # @return [String]
```
`@return [?]`

peterdavidhamilton commented on spec/helpers/gov_one_helper_spec.rb:
```diff
@@ -4,28 +4,36 @@
+  describe '#login_button' do
+    subject(:login_button) { helper.login_button }
+
+    it 'returns a link to the login uri' do
```
The description of this test is not actually what it checks.

jack-coggin replied:
[36ee42d](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/36ee42d12e27394f9c0a92bce46dc792fed49912)

peterdavidhamilton commented on spec/helpers/gov_one_helper_spec.rb:
```diff
@@ -4,28 +4,36 @@
-    it 'returns a URI object with the correct host and path' do
-      expect(logout_uri.host).to eq 'oidc.test.account.gov.uk'
-      expect(logout_uri.path).to eq '/logout'
+    it 'encodes the logout endpoint params' do
```
If you preferred having a URI object at any point you can move `to_s` into the button helpers.

jack-coggin replied:
@peterdavidhamilton I've updated the helper, was having trouble getting the login link to work - issue seems to be fixed by not encoding the redirect param

peterdavidhamilton replied:
Is that in local dev or when testing?

peterdavidhamilton commented on spec/models/user_spec.rb:
```diff
@@ -206,4 +206,39 @@
     end
   end
+
+  describe '.find_or_create_from_gov_one' do
```
See if you can improve readability using `with/without/and` rather than `when`

jack-coggin replied:
[36ee42d](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/36ee42d12e27394f9c0a92bce46dc792fed49912)

peterdavidhamilton commented on config/application.rb:
```diff
@@ -89,6 +96,11 @@ def debug?
     end
 
+    # @return [Boolean]
+    def gov_one_login?
```
Is this needed now?

jack-coggin replied:
No

peterdavidhamilton commented on spec/models/user_spec.rb:
```diff
@@ -206,4 +206,39 @@
     end
   end
+
+  describe '.find_or_create_from_gov_one' do
```
Again, is it at all possible to run prerequisite steps before assertions?

peterdavidhamilton commented on app/views/gov_one/info.html.slim:
```diff
@@ -0,0 +1,14 @@
+
+.govuk-grid-row
+  .govuk-grid-column-one-half
+    = m('gov-one-info.hero-body')
```
This is wrong again, there were locales before or am I mistaken. You don't want to be creating `resources` in the `test` env.

peterdavidhamilton commented on app/helpers/gov_one_helper.rb:
```diff
@@ -0,0 +1,49 @@
+  # @return [URI]
+  def logout_uri
+    params = {
+      post_logout_redirect_uri: GovOneAuthService::CALLBACKS[:logout],
```
I have no issue locally, so I am now going to retest in prod @jack-coggin

jack-coggin replied:
OK, will retest and revert

jack-coggin replied:
Right OK, thanks

peterdavidhamilton commented on app/services/gov_one_auth_service.rb:
```diff
@@ -0,0 +1,125 @@
+
+  # @param request [Net::HTTP::Get, Net::HTTP::Post]
+  # @return [Hash]
+  def handle_request(request)
```
The method `#handle_request` defined here is not used by any other code. Do you need it?

jack-coggin commented on spec/services/gov_one_auth_service_spec.rb:
```diff
@@ -0,0 +1,114 @@
+require 'rails_helper'
+
+RSpec.describe GovOneAuthService do
```
[3eb13d5](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/3eb13d5623bb4590cd6c8318a7b778510eb645df)

peterdavidhamilton commented:
https://github.com/DFE-Digital/early-years-foundation-recovery/pull/967#issuecomment-1846893221

peterdavidhamilton commented:
`0.001%` 😮

peterdavidhamilton commented:
@jack-coggin please copy the technical checklist requirements over to the PR description so they can be checked off

ebrett reviewed (approved):
lgtm

peterdavidhamilton commented on adr/ADR.md:
```diff
@@ -20,7 +20,8 @@ This log lists the architectural decisions for EYFS Recovery
 * [ADR-0014](0014-user-tracking.md) - Use Hotjar for tracking user journeys
 * [ADR-0015](0015-background-jobs.md) - Use Que for processing background jobs
-* [ADR-0016](0016-devise-security-and-pwned-passwords-gems.md) - Use Devise Security and Devise Pwned Password gems
+* [ADR-0016](0016-devise-security-and-pwned-password-gems.md) - Use gems Devise Security and Deviese Pwned Passwords
```
`Deviese`

jack-coggin replied:
[9bd4c17](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/9bd4c17db8db4c18a5ac02c14603efeb537fe85e)

martikat reviewed (approved):
LGTM 👍 I have run this locally and created a new account when checking this in the review app
